_posts/2020-05-07-health-groups.adoc

```diff
@@ -0,0 +1,235 @@
+= Health Groups - experimental MicroProfile Health extension in SmallRye
+
+If you are reading this blog, you have probably already heard about
+https://github.com/eclipse/microprofile-health[Eclipse MicroProfile Health]
+technology. In this article, we will introduce a new experimental extension to
+MicroProfile Health 2.2 called Health groups.
+
+All of the source code that we will present in this article is available at this
+GitHub repository https://github.com/xstefank/health-groups.
+
+== MicroProfile 2.2 health checks
+
+As you know, to define a health check procedure in MicroProfile Health you need
+something like this:
+
+[source,java]
+----
+package io.xstefank;
+
+import org.eclipse.microprofile.health.HealthCheck;
+import org.eclipse.microprofile.health.HealthCheckResponse;
+import org.eclipse.microprofile.health.Readiness;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped // 1. CDI scope
+@Readiness         // 2. Qualifier (Readiness, Liveness, Health)
+public class ReadinessHealthCheck implements HealthCheck {   // 3. implements HealthCheck interface
+    
+    @Override
+    public HealthCheckResponse call() {
+        return HealthCheckResponse.up(ReadinessHealthCheck.class.getSimpleName());
+    }
+}
+----
+
+So just to recap, to define a health check your class:
+
+- needs to be a CDI bean
+- needs to have one of the health qualifiers (Readiness, Liveness, or deprecated Health)
+- must implement the HealthCheck interface
+
+This is everything that you can use in MicroProfile 2.2. But what if you have a
+use case that doesn't fall into one of Readiness or Liveness categories? The answer
+is custom health groups.
+
+== Health Groups
+
+Quarkus 1.4.0 (with SmallRye Health 2.2.0) comes with a new addition to this API 
+called health groups. A health group is basically a custom group of health check 
+procedures that the user defines based on individual needs. As you probably noticed,
+Readiness and Liveness health checks already represent two predefined health groups. 
+With the health groups feature you can define as many additional health check groups
+as you need.
+
+=== Defining a health Group
+
+To define a health group you need only to declare at least one health check
+procedure that belongs to the health group:
+
+[source,java]
+----
+package io.xstefank;
+
+import io.smallrye.health.HealthGroup;
+import org.eclipse.microprofile.health.HealthCheck;
+import org.eclipse.microprofile.health.HealthCheckResponse;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+@HealthGroup("custom-group-1")
```

Comment:
I immediately think "String-based resolution is discouraged in CDI" :-)

You should be able to define your own health annotation, like

```java
@HealthGroup("my-health")
@Retention(RUNTIME)
@Qualifier
public @interface MyHealth {
}
```

and use that:

```java
@MyHealth
public class CustomHealthCheck implements HealthCheck {
    ...
}
```

Please take that as a feedback on the API, not on the article :-)

Reply:
Sure :) But how we could then get the custom qualifier when the `:8080/health/group/MyHealth` would be requested? Don't you need the concrete class of the qualifier for CDI select?

Also there is `@Named` in CDI right? I need to reread the spec.. :D

Reply:
Can't you remember, during initialization, what qualifier class corresponds to given health group name?

Reply:
I mean, I personally think this feature is more important for frameworks, rather than user applications. For example, I think one framework (can't remember which at the moment) defines a concept of wellness, so they would provide a `@Wellness` annotation that applications would use. The number of custom health groups would most likely be very low, I think.

Reply:
Maybe that could work. Definitely worth investigating. Wellness will be probably added on the same level as Readiness and Liveness to the spec. However, this feature was requested in the community. This is exactly the reason we can innovate in smallrye: We'll see and if it won't work we won't make it permanent right? :)

Reply:
Sure -- I'm providing feedback in that exact spirit :-)